### PR TITLE
connect: fix possible uninitialized memory access in getaddressinfo()

### PR DESCRIPTION
**Pull request by sstruchtrup**
I am using libcurl over anonymous/unbound AF_LOCAL sockets via socketpair() to test-drive some HTTP server code, using CURLOPT_OPENSOCKETFUNCTION to inject the file descriptor and CURLOPT_SOCKOPTFUNCTION to prevent a furter connect().

Admitting that this is somewhat of a corner-case.

On linux, sun_path in struct sockaddr_un is not initialized by accept(), getsockaddr() or getpeername() if the socket is not bound. This can only be detected by evaluating the length of the address, as returned by any of these functions.

The current implementation of getaddressinfo() in curl does not evaluate the address length and expects a null-terminated string in sun_path and therefore may access uninitialized data. getaddressinfo() won't write more than MAX_IPADR_LEN bytes, hence I believe no buffer overflow is possible.

I am not sure if it is worth fixing, but I am submitting this for review.

### CODE REVIEW

File: lib/connect.c [hunk 1]
```diff
@@ -660,7 +662,15 @@ UNITTEST bool getaddressinfo(struct sockaddr *sa, char *addr,
 #if defined(HAVE_SYS_UN_H) && defined(AF_UNIX)
     case AF_UNIX:
       su = (struct sockaddr_un*)sa;
-      msnprintf(addr, MAX_IPADR_LEN, "%s", su->sun_path);
+      /* On Linux, sun_path within struct sockaddr_un is not initialized
+         if an unbound AF_UNIX socket is returned from accept() or
+         getsockname()/getpeername() are invoked for an unbound socket.
+         Evaluate the address length to not access sun_path in this case. */
+      if(socklen > offsetof(struct sockaddr_un, sun_path))
+        msnprintf(addr, MAX_IPADR_LEN, "%s", su->sun_path);
```
Comment (bagder) [inline lib/connect.c]: This check puzzles me. You say `sun_path` ends up uninitialized here and that's the situation the condition here tries to avoid.

But how is checking if the storage struct is bigger than the offset to the `sun_path` field in the `sockaddr_un` a good check for this? This seems to rely on the representation of a struct that could vary and a destination buffer size that also could vary, to detect something that is unrelated to those two values.

Comment (sstruchtrup) [reply lib/connect.c]: The variable is actually misnamed here, only seeing that now. `addrlen` oder `salen` would be better here. `accept()` and `getpeername()`/`getsockname()` return (via a pointer) the length of the address, i.e. the amount of initialized data, which is usually equal to the size of the corresponding socket adress struct for that address family, but not so for AF_LOCAL, where the struct is partially initialized only, due to the variable length `sun_path`.

On my linux system, getsockname() would return 2 for an unbound AF_LOCAL socket.

Taking the definition of `struct sockaddr_un`:
```
struct sockaddr_un {
        __kernel_sa_family_t sun_family; /* AF_UNIX */
        char sun_path[UNIX_PATH_MAX];   /* pathname */
};
```
Hence the address length of 2 indicates that only `sun_family` but not `sun_path` are initialized.

**Comment by sstruchtrup** [comment]
A quick example demonstrating the problem:

```C
#include <stdio.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>


int main()
{
        int fds[2];
        if (socketpair(AF_LOCAL, SOCK_STREAM, 0, fds) == -1)
        {
                perror("socketpair");
                return 1;
        }

        struct sockaddr_un sa;
        // Initialize sa.sun_path
        strcpy(sa.sun_path, "<uninitialized>");

        socklen_t salen;
        if (getsockname(fds[0], (struct sockaddr *)&sa, &salen) == -1)
        {
                perror("getsockname");
                return 1;
        }

        fprintf(stderr, "salen: %u\n", salen);
        fprintf(stderr, "sun_path: %s\n", sa.sun_path);

        return 0;
}
```

The output is:
```
2
<unitialized>
```

**Comment by bagder** [comment]
Yes, that's true. But we (should) clear the struct first before calling `getsockname`, like the diff below, which then makes this output:
 
    $ ./a.out 
    salen: 2
    sun_path: 

~~~diff
--- saclean.c~  2019-05-27 10:38:40.826213171 +0200
+++ saclean.c   2019-05-27 10:39:10.306436212 +0200
@@ -18,6 +18,7 @@
         strcpy(sa.sun_path, "<uninitialized>");
 
         socklen_t salen;
+        memset(&sa, 0, sizeof(sa));
         if (getsockname(fds[0], (struct sockaddr *)&sa, &salen) == -1)
         {
                 perror("getsockname");

**Comment by bagder** [comment]
... which I now realize just shows the problem:
~~~
$ valgrind ./a.out 
==18540== Memcheck, a memory error detector
==18540== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==18540== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==18540== Command: ./a.out
==18540== 
==18540== Syscall param socketcall.getsockname(namelen_in) points to uninitialised byte(s)
==18540==    at 0x497C597: getsockname (syscall-template.S:78)
==18540==    by 0x109206: main (in /home/daniel/tmp/a.out)
==18540==  Address 0x1ffefffd2c is on thread 1's stack
==18540==  in frame #1, created by main (???:)
==18540== 
==18540== Syscall param socketcall.getsockname(namelen_out) points to uninitialised byte(s)
==18540==    at 0x497C597: getsockname (syscall-template.S:78)
==18540==    by 0x109206: main (in /home/daniel/tmp/a.out)
==18540==  Address 0x1ffefffd2c is on thread 1's stack
==18540==  in frame #1, created by main (???:)
==18540== 
==18540== Use of uninitialised value of size 8
==18540==    at 0x48CED1E: _itoa_word (_itoa.c:179)
==18540==    by 0x48D25F3: vfprintf (vfprintf.c:1637)
==18540==    by 0x48D4865: buffered_vfprintf (vfprintf.c:2322)
==18540==    by 0x48D1EB1: vfprintf (vfprintf.c:1296)
==18540==    by 0x48DA533: fprintf (fprintf.c:32)
==18540==    by 0x10923F: main (in /home/daniel/tmp/a.out)
==18540== 
==18540== Conditional jump or move depends on uninitialised value(s)
==18540==    at 0x48CED29: _itoa_word (_itoa.c:179)
==18540==    by 0x48D25F3: vfprintf (vfprintf.c:1637)
==18540==    by 0x48D4865: buffered_vfprintf (vfprintf.c:2322)
==18540==    by 0x48D1EB1: vfprintf (vfprintf.c:1296)
==18540==    by 0x48DA533: fprintf (fprintf.c:32)
==18540==    by 0x10923F: main (in /home/daniel/tmp/a.out)
==18540== 
==18540== Conditional jump or move depends on uninitialised value(s)
==18540==    at 0x48D3213: vfprintf (vfprintf.c:1637)
==18540==    by 0x48D4865: buffered_vfprintf (vfprintf.c:2322)
==18540==    by 0x48D1EB1: vfprintf (vfprintf.c:1296)
==18540==    by 0x48DA533: fprintf (fprintf.c:32)
==18540==    by 0x10923F: main (in /home/daniel/tmp/a.out)
==18540== 
==18540== Conditional jump or move depends on uninitialised value(s)
==18540==    at 0x48D275D: vfprintf (vfprintf.c:1637)
==18540==    by 0x48D4865: buffered_vfprintf (vfprintf.c:2322)
==18540==    by 0x48D1EB1: vfprintf (vfprintf.c:1296)
==18540==    by 0x48DA533: fprintf (fprintf.c:32)
==18540==    by 0x10923F: main (in /home/daniel/tmp/a.out)
==18540== 
salen: 2
sun_path: 
==18540== 
==18540== HEAP SUMMARY:
==18540==     in use at exit: 0 bytes in 0 blocks
==18540==   total heap usage: 0 allocs, 0 frees, 0 bytes allocated
==18540== 
==18540== All heap blocks were freed -- no leaks are possible
==18540== 
==18540== For counts of detected and suppressed errors, rerun with: -v
==18540== Use --track-origins=yes to see where uninitialised values come from
==18540== ERROR SUMMARY: 6 errors from 6 contexts (suppressed: 0 from 0)

**Comment by bagder** [comment]
The example was wrong. It didn't call `getsockname` correctly. A fixed example that clears the struct and get initialized memory back:

~~~c
#include <stdio.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>


int main()
{
        int fds[2];
        if (socketpair(AF_LOCAL, SOCK_STREAM, 0, fds) == -1)
        {
                perror("socketpair");
                return 1;
        }

        struct sockaddr_storage sa;
        struct sockaddr_un *un = (struct sockaddr_un *)&sa;
        socklen_t salen = sizeof(sa);
        memset(&sa, 0, sizeof(sa));
        fprintf(stderr, "salen in: %u\n", salen);
        if (getsockname(fds[0], (struct sockaddr *)&sa, &salen) == -1)
        {
                perror("getsockname");
                return 1;
        }

        fprintf(stderr, "salen: %u\n", salen);
        fprintf(stderr, "sun_path: %s\n", un->sun_path);

        return 0;
}

**Comment by sstruchtrup** [comment]
Would you be willing to commit a memset() to clear the struct sockaddr as per your proposal?

Taken that this isn't necessary for (nearly) all ordinary use cases.

**Comment by bagder** [comment]
Yes, such a `memset()` call should be done before calling `getsockname()`. You up to making a PR for whatever call that is causing you this problem?

**Comment by sstruchtrup** [comment]
Sure, I'll do that. Just need a few days.
